File: leet-code/198. House Robber/HouseRobber198_2.java
package com.example.algorithm;

public class HouseRobber198_2 {

    public int rob(int[] nums) {
        int[] dp = new int[nums.length];
        if (nums.length == 1) return nums[0];
        dp[0] = nums[0];
        dp[1] = Math.max(dp[0], nums[1]);
        for (int i = 2; i < nums.length; i++)
            dp[i] = Math.max(dp[i - 1], dp[i - 2] + nums[i]);
        return dp[nums.length - 1];
    }
}

File: leet-code/198. House Robber/HouseRobber198_2.kt
package com.example

import kotlin.math.max

class HouseRobber198_2 {
    fun rob(nums: IntArray): Int {
        // dp
        // 직전의 값을 그대로 올리거나
        // 직전 이전의 값과 더하거나
        if (nums.size == 1) {
            return nums[0]
        }
        nums[1] = max(nums[0], nums[1])
        for (i in 2..nums.lastIndex) {
            nums[i] = max(nums[i - 1], nums[i - 2] + nums[i])
        }
        return nums.last()
    }
}
